include/tvm/topi/einsum.h

```diff
@@ -0,0 +1,930 @@
+                           iter_labels),
+             0);
+    for (idim = 0; idim < ndim_iter; idim++) {
+      if (op_axes[iop][idim] != -1) {
+        iterstride[iop].Set(idim, opstride_true[iop][op_axes[iop][idim]]);
+        if (GetConstInt(itershape[idim]) != -1) {
+          if (GetConstInt(itershape[idim]) == 1) {
+            itershape.Set(idim, opshape[iop][op_axes[iop][idim]]);
+          }
+        } else {
+          itershape.Set(idim, opshape[iop][op_axes[iop][idim]]);
+        }
+      }
+    }
+  }
+  for (idim = 0; idim < ndim_output; ++idim) {
+    iterstride[nop].Set(idim, ostride_true[idim]);
+  }
+  reduceshape = Array<PrimExpr>(static_cast<size_t>(ndim_iter - ndim_output), 0);
+  for (idim = ndim_output; idim < ndim_iter; ++idim) {
+    reduceshape.Set(idim - ndim_output, itershape[idim]);
+  }
+  for (iop = 0; iop < nop; iop++) {
+    Array<Integer> rsh;
+    for (idim = 0; idim < ndim_iter; idim++) {
+      if (op_axes_arrays[iop][idim] == -1) {
+        rsh.push_back(GetConstInt(itershape[idim]));
+      } else {
+        if (GetConstInt(itershape[idim] != opshape[iop][op_axes_arrays[iop][idim]])) {
+          rsh.push_back(GetConstInt(itershape[idim]));
+        }
+      }
+    }
+    remainshape[iop] = Array<PrimExpr>(rsh.begin(), rsh.end());
+  }
+  // exclude the 0-dim case
+  if (ndim_iter == 0) {
+    ndim_iter = 1;
+  }
+  itershape = Pad(itershape, ndim_iter);
+  for (iop = 0; iop <= nop; ++iop) {
+    iterstride[iop] = Pad(iterstride[iop], ndim_iter);
+  }
+  // oshape = Pad(oshape, ndim_iter);
+  reduceshape = Pad(reduceshape, ndim_iter);
+  for (iop = 0; iop < nop; ++iop) {
+    opshape[iop] = Pad(opshape[iop], ndim_iter);
+    remainshape[iop] = Pad(remainshape[iop], ndim_iter);
+  }
+  // ostride and rstride
+  Array<Array<PrimExpr>> ostride;
+  Array<Array<PrimExpr>> rstride;
+
+  for (iop = 0; iop < nop; ++iop) {
+    Array<PrimExpr> otmp(static_cast<size_t>(ndim_iter), 0);
+    Array<PrimExpr> rtmp(static_cast<size_t>(ndim_iter), 0);
+    for (idim = 0; idim < ndim_iter; ++idim) {
+      otmp.Set(idim, idim < ndim_output ? iterstride[iop][idim] : 1);
+      rtmp.Set(idim, idim < ndim_iter - ndim_output ? iterstride[iop][idim + ndim_output] : 1);
+    }
+    ostride.push_back(otmp);
+    rstride.push_back(rtmp);
+  }
+
+  // func: input indices => return cooresponding value
+  auto func = [inputs, oshape, ostride, reduceshape, ndim_iter, rstride,
+               nop](const Array<Var>& input_indices) -> PrimExpr {
+    for (int rdim = 0; rdim < ndim_iter; ++rdim) {
+      if (GetConstInt(reduceshape[rdim]) == 0) {
+        return 0;  //
+      }
+    }
+    Array<PrimExpr> ridx = UnravelIndex(0, reduceshape);
+    // AType needs to fit the output data
+    PrimExpr sum = 0;
+    bool rec_flag = false;
+    do {
+      PrimExpr tmp = 1;
+      for (int iop = 0; iop < nop; ++iop) {
+        if (iop != -1) {
+          PrimExpr k = 0;
+          // k = dot(input_indices, ostride) + dot(ridx, rstride)
```

Comment:
is this comment what following code tries to implement?

Reply:
Yeah. Already removed. Thx!